tortoisebeacon/tortoise_beacon.go

```diff
@@ -0,0 +1,917 @@
+		potentiallyValidProposals:       make(map[types.EpochID]hashSet),
+		ownVotes:                        make(ownVotes),
+		beacons:                         make(map[types.EpochID]types.Hash32),
+		proposalPhaseFinishedTimestamps: map[types.EpochID]time.Time{},
+		incomingVotes:                   map[epochRoundPair]votesPerPK{},
+		firstRoundIncomingVotes:         map[types.EpochID]firstRoundVotesPerPK{},
+		firstRoundOutcomingVotes:        map[types.EpochID]firstRoundVotes{},
+		seenEpochs:                      map[types.EpochID]struct{}{},
+	}
+}
+
+// Start starts listening for layers and outputs.
+func (tb *TortoiseBeacon) Start(ctx context.Context) error {
+	tb.Log.Info("Starting %v with the following config: %+v", protoName, tb.config)
+
+	tb.initGenesisBeacons()
+
+	tb.layerTicker = tb.clock.Subscribe()
+
+	tb.backgroundWG.Add(1)
+
+	go func() {
+		defer tb.backgroundWG.Done()
+
+		tb.listenLayers(ctx)
+	}()
+
+	tb.backgroundWG.Add(1)
+
+	go func() {
+		defer tb.backgroundWG.Done()
+
+		tb.cleanupLoop()
+	}()
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) initGenesisBeacons() {
+	closedCh := make(chan struct{})
+	close(closedCh)
+
+	epoch := types.EpochID(0)
+	for ; epoch.IsGenesis(); epoch++ {
+		tb.beacons[epoch] = genesisBeacon
+	}
+}
+
+// Close closes TortoiseBeacon.
+func (tb *TortoiseBeacon) Close() error {
+	tb.Log.Info("Closing %v", protoName)
+	tb.Closer.Close()
+	tb.backgroundWG.Wait() // Wait until background goroutines finish
+	tb.clock.Unsubscribe(tb.layerTicker)
+
+	return nil
+}
+
+// GetBeacon returns a Tortoise Beacon value as []byte for a certain epoch or an error if it doesn't exist.
+func (tb *TortoiseBeacon) GetBeacon(epochID types.EpochID) ([]byte, error) {
+	if tb.tortoiseBeaconDB != nil {
+		if val, ok := tb.tortoiseBeaconDB.GetTortoiseBeacon(epochID); ok {
+			return val.Bytes(), nil
+		}
+	}
+
+	if (epochID - 1).IsGenesis() {
+		return genesisBeacon.Bytes(), nil
+	}
+
+	tb.beaconsMu.RLock()
+	defer tb.beaconsMu.RUnlock()
+
+	beacon, ok := tb.beacons[epochID-1]
+	if !ok {
+		tb.Log.With().Error("Beacon is not calculated",
+			log.Uint64("target_epoch", uint64(epochID)),
+			log.Uint64("beacon_epoch", uint64(epochID-1)))
+
+		return nil, ErrBeaconNotCalculated
+	}
+
+	if tb.tortoiseBeaconDB != nil {
+		if err := tb.tortoiseBeaconDB.SetTortoiseBeacon(epochID, beacon); err != nil {
+			return nil, fmt.Errorf("update beacon in DB: %w", err)
+		}
+	}
+
+	return beacon.Bytes(), nil
+}
+
+func (tb *TortoiseBeacon) cleanupLoop() {
+	ticker := time.NewTicker(cleanupInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-tb.CloseChannel():
+			return
+		case <-ticker.C:
+			tb.cleanup()
+		}
+	}
+}
+
+func (tb *TortoiseBeacon) cleanup() {
+	// TODO(nkryuchkov): implement a better solution, consider https://github.com/golang/go/issues/20135
+	tb.beaconsMu.Lock()
+	defer tb.beaconsMu.Unlock()
+
+	for e := range tb.beacons {
+		if tb.epochIsOutdated(e) {
```

Comment:
I'd do it when inserting a new beacon - this will eliminate the need for an additional gouroutine and simply follow a LRU heuristic